compiler/rustc_codegen_ssa/src/traits/type_.rs

```diff
@@ -126,6 +126,28 @@ pub trait LayoutTypeMethods<'tcx>: Backend<'tcx> {
         index: usize,
         immediate: bool,
     ) -> Self::Type;
+
+    /// A type that can be used in a [`super::BuilderMethods::load`] +
+    /// [`super::BuilderMethods::store`] pair to implement a *typed* copy,
+    /// such as a MIR `*_0 = *_1`.
+    ///
+    /// It's always legal to return `None` here, as the provided impl does,
+    /// in which case callers should use [`super::BuilderMethods::memcpy`]
+    /// instead of the `load`+`store` pair.
+    ///
+    /// This can be helpful for things like arrays, where the LLVM backend type
+    /// `[3 x i16]` optimizes to three separate loads and stores, but it can
+    /// instead be copied via an `i48` that stays as the single `load`+`store`.
+    /// (As of 2023-05 LLVM cannot necessarily optimize away a `memcpy` in these
+    /// cases, due to `poison` handling, but in codegen we have more information
+    /// about the type invariants, so can emit something better instead.)
+    ///
+    /// This *should* return `None` for particularly-large types, where leaving
+    /// the `memcpy` may well be important to avoid code size explosion.
+    fn scalar_copy_backend_type(&self, layout: TyAndLayout<'tcx>) -> Option<Self::Type> {
```

Comment:
Having a default impl here may make it not obvious that other backends should emit typed load/store. Is it bad style to just add this to cranelift and codegen_gcc here too?

Reply:
Well, cg_clif doesn't use cg_ssa, so it's not impacted.

I have no practical way to test cg_gcc, being on windows, and I also don't actually have any information that doing something other than `memcpy` for it would actually be an improvement for it, so I figured I'd just leave the default here since it's a semantically *sufficient* implementation.

Notably, GCC apparently doesn't have the `poison` semantics that are what nikic mentioned as being the problem for better optimizing this:

https://github.com/rust-lang/rust/blob/f8447b963893b410f1dbce2cd752a898dc3ed8d4/compiler/rustc_codegen_gcc/src/common.rs#L76-L79

so indeed it might just never need to do this.